pkg/agent/controller/networkpolicy/fqdn.go

```diff
@@ -0,0 +1,782 @@
+	if !ok {
+		f.fqdnToSelectorItem[fqdn] = map[fqdnSelectorItem]struct{}{
+			selectorItem: {},
+		}
+	} else {
+		matchedSelectorItems[selectorItem] = struct{}{}
+	}
+	matchedFQDNs, ok := f.selectorItemToFQDN[selectorItem]
+	if !ok {
+		f.selectorItemToFQDN[selectorItem] = sets.NewString(fqdn)
+	} else {
+		matchedFQDNs.Insert(fqdn)
+	}
+}
+
+// getIPsForFQDNSelectors retrieves the current IP addresses cached for FQDNs that
+// matches the selection criteria of a v1beta2.FQDN selector.
+func (f *fqdnController) getIPsForFQDNSelectors(fqdns []string) []net.IP {
+	f.fqdnSelectorMutex.Lock()
+	defer f.fqdnSelectorMutex.Unlock()
+	var matchedIPs []net.IP
+	for _, fqdn := range fqdns {
+		fqdnSelectorItem := fqdnToSelectorItem(fqdn)
+		fqdnsMatched, ok := f.selectorItemToFQDN[fqdnSelectorItem]
+		if !ok {
+			klog.Errorf("FQDN selector %v is not known to the controller, cannot get IPs", fqdnSelectorItem)
+			return matchedIPs
+		}
+		for fqdn := range fqdnsMatched {
+			if dnsMeta, ok := f.dnsEntryCache[fqdn]; ok {
+				for _, ip := range dnsMeta.responseIPs {
+					matchedIPs = append(matchedIPs, ip)
+				}
+			}
+		}
+	}
+	return matchedIPs
+}
+
+// addFQDNRule adds a new FQDN rule to fqdnSelectorItem mapping, as well as the OFAddresses of
+// Pods selected by the FQDN rule.
+func (f *fqdnController) addFQDNRule(ruleID string, fqdns []string, podOFAddrs sets.Int32) error {
+	f.addFQDNSelector(ruleID, fqdns)
+	return f.updateRuleSelectedPods(ruleID, podOFAddrs)
+}
+
+func (f *fqdnController) addFQDNSelector(ruleID string, fqdns []string) {
+	f.fqdnSelectorMutex.Lock()
+	defer f.fqdnSelectorMutex.Unlock()
+	for _, fqdn := range fqdns {
+		fqdnSelectorItem := fqdnToSelectorItem(fqdn)
+		ruleIDs, exists := f.selectorItemToRuleIDs[fqdnSelectorItem]
+		if !exists {
+			// This is a new fqdnSelectorItem. All existing FQDNs in the cache needs to be matched
+			// against this fqdnSelectorItem to update the mapping.
+			f.selectorItemToRuleIDs[fqdnSelectorItem] = sets.NewString(ruleID)
+			for fqdn := range f.dnsEntryCache {
+				if fqdnSelectorItem.matches(fqdn) {
+					f.setFQDNMatchSelector(fqdn, fqdnSelectorItem)
+				}
+			}
+		} else {
+			f.selectorItemToRuleIDs[fqdnSelectorItem] = ruleIDs.Insert(ruleID)
+		}
+		if fqdnSelectorItem.matchName != "" {
+			// Start a DNS query immediately for matchName selectors.
+			f.setFQDNMatchSelector(fqdnSelectorItem.matchName, fqdnSelectorItem)
+			f.dnsQueryQueue.Add(fqdnSelectorItem.matchName)
+		}
+	}
+}
+
+// updateRuleSelectedPods updates the Pod OFAddresses selected by a FQDN rule. Those addresses
+// are used to create DNS response interception rules.
+func (f *fqdnController) updateRuleSelectedPods(ruleID string, podOFAddrs sets.Int32) error {
+	f.fqdnRuleToPodsMutex.Lock()
+	defer f.fqdnRuleToPodsMutex.Unlock()
+	originalPodSet, newPodSet := sets.Int32{}, sets.Int32{}
+	for _, pods := range f.fqdnRuleToSelectedPods {
+		originalPodSet = originalPodSet.Union(pods)
```

Comment:
done, I chose to rename the functions to `MergeInt32` and `MergeString` rather than have different sub-packages. Happy to change that if you prefer the latter, I don't have a strong preference myself.